src/backend/storage/data_table.cpp

```diff
@@ -1009,5 +1018,468 @@ column_map_type DataTable::GetStaticColumnMap(const std::string &table_name,
+ */
+size_t DataTable::SampleRows(size_t sample_size) {
+  LOG_INFO("Start a new sampling, size = %lu ", sample_size);
+
+  // This function should be mutual exclusive
+  sample_mutex.lock();
+
+  if(samples_for_optimizer.size() != 0) {
+    LOG_INFO("Previous sample size not 0. Clear! ");
+
+    // Clear previous sample results, if there is one
+    samples_for_optimizer.clear();
+  }
+
+  // If there is an existing materialized sample tile group
+  // drop it first to avoid data inconsistency
+  if(sampled_tile_group_id != INVALID_OID) {
+    LOG_INFO("Dropping an old sampled tile group... Prepare a new one.");
+
+    catalog::Manager::GetInstance().DropTileGroup(sampled_tile_group_id);
+  }
+
+  // If there is old information in cardinality map also clear them
+  if(cardinality_map.size() != 0) {
+    LOG_INFO("Clearing existing cardinality map..\n");
+
+    cardinality_map.clear();
+  }
+
+  // We want to make it ordered, and also want to detect for duplicates
+  // so use a RB-Tree based ordered set
+  std::set<oid_t> row_id_set{};
+
+  // Get tile group count and tuple count
+  // This is different from number of tuple which is just a float point
+  // approximation - this is exact accurate number
+  size_t total_tuple_number = tuple_count_exact;
+
+  if(sample_size >= total_tuple_number) {
+    LOG_INFO("Sample size too large! Adjust to fit actual table size %lu... ",
+              total_tuple_number);
+
+    sample_size = total_tuple_number;
+
+    // Do not use random generator if we know we will cover all rows
+    // Just insert them in increasing order
+    for(size_t i = 0;i < sample_size;i++) {
+      row_id_set.insert(i);
+    }
+  } else {
+    // I copied this from the Internet...
+    std::random_device rd;
+    std::mt19937 generator(rd());
+
+    // This is inclusive, [0, total_tuple_number - 1]
+    std::uniform_int_distribution<> distribution(0, total_tuple_number - 1);
+
+    int iter_count = 0;
+    peloton::concurrency::TransactionManager &txn_mgr = \
+      peloton::concurrency::TransactionManagerFactory::GetInstance();
+
+    // outer loop detect whether there are duplicates
+    while((row_id_set.size() < sample_size) && \
+          (iter_count < 10)) {
+      // Inner loop generates random numbers and insert into
+      // the ordered set
+      for(size_t i = 0;i < sample_size;i++) {
+        oid_t random_row_id = distribution(generator);
+        // This is the tile group ID inside the vector
+        oid_t tile_group_offset = random_row_id / tuples_per_tilegroup;
+        auto tile_group_p = GetTileGroup(tile_group_offset);
+
+        // We need this header to check whether tuple is visible to
+        // current thread
+        TileGroupHeader *header_p = tile_group_p->GetHeader();
+
+        // According to txn manager, this tuple is visible to current txn
+        // so we take it as the sample
+        if(txn_mgr.IsVisible(header_p, random_row_id)) {
+          row_id_set.insert(random_row_id);
```

Comment:
it may unsafe to reuse sampled itempointer, because 1) though at this point, this version is visible to the current transactions. Later on, this version could be stale, or an aborted version.  Even worse, if gc is enabled, this itempointer may be recycled and become useless.